MLP.ipynb

```diff
@@ -1,1146 +1,1188 @@
+    "            y_val_batch = y_val_batch.to(device)\n",
+    "\n",
+    "            # Prepare labels and input\n",
+    "            y_val = F.one_hot(y_val_batch, num_classes=nc).type(torch.float32)\n",
+    "\n",
+    "            # Forward pass through CNN and fully connected network\n",
+    "            cnn_output_val = model(x_val_batch)\n",
+    "            cnn_output_np_val = cnn_output_val.cpu().detach().numpy()\n",
+    "            cnn_output_np_val = nn_network.flatten(cnn_output_np_val)\n",
+    "            yhat_val = nn_network.feedforward(cnn_output_np_val)\n",
+    "\n",
+    "            # Compute validation loss\n",
+    "            y_val_np = y_val.cpu().detach().numpy()\n",
+    "            val_loss = nn_network.compute_loss(yhat_val, y_val_np)\n",
+    "            total_val_loss += val_loss\n",
+    "\n",
+    "            # Calculate validation accuracy\n",
+    "            val_preds = np.argmax(yhat_val, axis=1)\n",
+    "            correct_val_preds += np.sum(val_preds == np.argmax(y_val_np, axis=1))\n",
+    "            total_val_samples += y_val_batch.size(0)\n",
+    "\n",
+    "    # Calculate average validation loss and accuracy\n",
+    "    val_loss = total_val_loss / len(val_loader)\n",
+    "    val_accuracy = correct_val_preds / total_val_samples\n",
+    "\n",
+    "    return val_loss, val_accuracy"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_and_validate(model, train_loader, val_loader, nn_network, device, nc, learning_rate=0.0001, epochs=1):\n",
+    "    train_accuracies, train_losses, val_accuracies, val_losses = [], [], [], []\n",
+    "\n",
+    "    for epoch in range(epochs):\n",
+    "        print(f\"Epoch {epoch + 1}/{epochs}\")\n",
+    "\n",
+    "        # Train the model for one epoch\n",
+    "        train_loss, train_accuracy = train_one_epoch(model, train_loader, nn_network, device, nc, learning_rate)\n",
+    "        train_losses.append(train_loss)\n",
+    "        train_accuracies.append(train_accuracy)\n",
+    "        print(f\"Train loss: {train_loss:.4f}, Train accuracy: {train_accuracy:.4f}\")\n",
+    "\n",
+    "        # Validate the model\n",
+    "        val_loss, val_accuracy = validate(model, val_loader, nn_network, device, nc)\n",
+    "        val_losses.append(val_loss)\n",
+    "        val_accuracies.append(val_accuracy)\n",
+    "        print(f\"Val loss: {val_loss:.4f}, Val accuracy: {val_accuracy:.4f}\")\n",
+    "\n",
+    "    return train_accuracies, train_losses, val_accuracies, val_losses"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_accuracies, train_losses, val_accuracies, val_losses = train_and_validate(\n",
+    "    model, train_loader, val_loader, nn_network, device, nc, learning_rate=0.0001, epochs=1\n",
```

Comment:
NIT: can we keep it in dict or as a variable so that functions are not disturbed.